ttnn/cpp/ttnn/tensor/tensor.cpp

```diff
@@ -734,32 +735,21 @@ Tensor create_device_tensor(
 namespace detail {
 template <typename DataType>
 void* get_raw_host_data_ptr(const Tensor& tensor) {
-    return std::visit(
-        [](auto&& storage) -> void* {
-            using StorageType = std::decay_t<decltype(storage)>;
-            if constexpr (std::is_same_v<StorageType, OwnedStorage>) {
-                auto buffer = owned_buffer::get_as<DataType>(storage.buffer);
-                return buffer.data();
-            } else if constexpr (std::is_same_v<StorageType, BorrowedStorage>) {
-                if constexpr (
-                    std::is_same_v<DataType, float> or std::is_same_v<DataType, bfloat16> or
-                    std::is_same_v<DataType, std::uint32_t> or std::is_same_v<DataType, std::int32_t> or
-                    std::is_same_v<DataType, std::uint8_t> or std::is_same_v<DataType, std::uint16_t>) {
-                    auto buffer = borrowed_buffer::get_as<DataType>(storage.buffer);
-                    return buffer.data();
-                } else {
-                    TT_THROW("Borrowed storage doesn't support this data type");
-                }
-            } else if constexpr (std::is_same_v<StorageType, DeviceStorage>) {
-                TT_THROW("Device storage isn't supported");
-            } else if constexpr (std::is_same_v<StorageType, MultiDeviceStorage>) {
-                TT_THROW("Device storage isn't supported");
-            } else if constexpr (std::is_same_v<StorageType, MultiDeviceHostStorage>) {
-                TT_THROW("Device storage isn't supported");
-            } else {
-                raise_unsupported_storage<StorageType>();
-            }
+    return std::visit(tt::stl::overloaded{
+        [](OwnedStorage& s) {
+            auto buffer = owned_buffer::get_as<DataType>(s.buffer);
+            return buffer.data();
+        },
+        [](BorrowedStorage& s)
+            requires(std::is_same_v<DataType, float> or std::is_same_v<DataType, bfloat16> or
+                     std::is_same_v<DataType, std::uint32_t> or std::is_same_v<DataType, std::int32_t> or
+                     std::is_same_v<DataType, std::uint8_t> or std::is_same_v<DataType, std::uint16_t>)
+        {
+            auto buffer = borrowed_buffer::get_as<DataType>(s.buffer);
+            return buffer.data();
         },
+        [](BorrowedStorage&) -> void* { TT_THROW("Borrowed storage doesn't support this data type"); },
```

Comment:
Are you sure you want to pass  BorrowedStorage& here?

Reply:
So this is actually compiler dependent - clang 17 happens to work, but this fails at higher versions. gcc fails up until 14. https://godbolt.org/z/Mnvnzs5Mr.

I reverted this back to `if constexpr`, seems like the simplest solution at this point.